transport_plugins/jlink/iotile_transport_jlink/jlink_background.py

```diff
@@ -262,76 +188,142 @@ def _debug_read_memory(self, device_info, _control_info, args, _progress_callbac
+                pc_reg, ff_cfg.ff_addresses['program_start'])
 
-            self._continue()
-            max_read_length = self._get_ff_max_read_length()
+            await self._continue()
+            max_read_length = await self._get_ff_max_read_length()
 
             bytes_dumped = 0
             memory = b''
             while bytes_dumped < data_length:
                 bytes_to_dump = max_read_length if (data_length - bytes_dumped) > max_read_length else (data_length - bytes_dumped)
- 
+
                 logger.info("BKPT hit, writing SPI dump command now...")
-                logger.info("At PC: %s", hex(self._jlink.register_read(pc_reg)))
-                self._write_ff_dump_cmd(start_addr, data_length)
+                logger.info("At PC: %s", hex(await self.register_read(pc_reg)))
+                await self._write_ff_dump_cmd(start_addr, data_length)
 
-                self._continue()
+                await self._continue()
 
                 logger.info("BKPT hit, reading response buffer address...")
-                logger.info("At PC: %s", hex(self._jlink.register_read(pc_reg)))
-                memory += self._read_ff_dump_resp(data_length)
+                logger.info("At PC: %s", hex(await self.register_read(pc_reg)))
+                memory += await self._read_ff_dump_resp(data_length)
 
                 bytes_dumped += bytes_to_dump
-                self._continue()
+                await self._continue()
 
-            self._jlink.reset()
-            self._jlink._dll.JLINKARM_Go()
+            await self.reset()
+            await self._loop.run_in_executor(self._jlink._dll.JLINKARM_Go)
 
         elif memory_region == 'flash' or memory_region == 'ram' or memory_region == 'mapped':
             if pause is True:
-                self._jlink.halt()
-            
-            memory = self._read_mapped_memory(device_info, memory_region, start_addr, data_length)
+                await self._loop.run_in_executor(self._jlink.halt)
+
+            memory = await self._read_mapped_memory(device_info, memory_region, start_addr, data_length)
 
             if pause is True:
-                self._jlink._dll.JLINKARM_Go()
+                await self._loop.run_in_executor(self._jlink._dll.JLINKARM_Go)
         else:
             raise ArgumentError("Invalid memory region specified, must be one of ['flash', 'ram', 'external', 'mapped']")
 
         return memory
 
-    def _debug_write_memory(self, _device_info, _control_info, args, _progress_callback):
+    async def debug_write_memory(self, _device_info, _control_info, args):
         memory_region = args.get('region').lower()
 
         if memory_region == 'external' or memory_region == 'flash':
             raise ArgumentError("Writing to flash/external is not currently supported.")
-        self._jlink.memory_write(args.get('address'), args.get('data'))
+        await self._loop.run_in_executor(
+            self._jlink.memory_write, args.get('address'), args.get('data'))
```

Comment:
Can't this just use the `async write_memory` method that you define below rather than running in an executor thread?